src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV4.cs

```diff
@@ -96,6 +79,52 @@ public override async Task<long> CreateInstanceIndexAsync(DicomDataset instance,
             }
         }
 
+        public override async Task EndCreateInstanceIndexAsync(
+            DicomDataset instance,
+            long watermark,
+            IEnumerable<QueryTag> queryTags,
+            CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNull(instance, nameof(instance));
+            EnsureArg.IsNotNull(queryTags, nameof(queryTags));
+
+            // Use maxTagVersion to track tag addition -- if new tag is added, max tag version increases.
+            var rows = ExtendedQueryTagDataRowsBuilder.Build(instance, queryTags);
+            VLatest.EndAddInstanceTableValuedParameters parameters = new VLatest.EndAddInstanceTableValuedParameters(
+                rows.StringRows,
+                rows.LongRows,
+                rows.DoubleRows,
+                rows.DateTimeRows,
+                rows.PersonNameRows);
+
+            using (SqlConnectionWrapper sqlConnectionWrapper = await SqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.EndAddInstance.PopulateCommand(
+                    sqlCommandWrapper,
+                    instance.GetSingleValueOrDefault(DicomTag.StudyInstanceUID, string.Empty),
+                    instance.GetSingleValueOrDefault(DicomTag.SeriesInstanceUID, string.Empty),
+                    instance.GetSingleValueOrDefault(DicomTag.SOPInstanceUID, string.Empty),
+                    watermark,
+                    UlongToRowVersion(rows.MaxVersion),
```

Comment:
I'm starting to wonder if we really need this field or whether we can just use `TagKey`?